webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/pdf/FmcSolutionSheet.kt

```diff
@@ -239,118 +250,109 @@ open class FmcSolutionSheet(request: ScrambleRequest, globalTitle: String?, loca
         for (i in directionModifiers.indices) {
+            val list = ArrayList<String>()
             for (j in rotations.indices) {
                 if (!rotations[j].isBlank()) {
-                    movesCell[1][i][j] = rotations[j].toLowerCase() + directionModifiers[i]
+                    list.add(rotations[j].toLowerCase() + directionModifiers[i])
                 } else {
-                    movesCell[1][i][j] = "";
+                    list.add("")
                 }
             }
+            rotationsList.add(list)
         }
+        movesCell.add(rotationsList)
+
+        // Table
+        val tableWidth = competitorInfoLeft - left - 2 * margin
+        val tableHeight = movesRect.height
+        val tableLines = movesType.size * (directionModifiers.size + 1) // F, F2, F' (direction modifiers) and the title
+        val cellWidth = 25f // TODO calculate
+        val cellHeight = tableHeight / tableLines
+        val columns = moves.size + 1 // R, U, F, L, D, B and the name
+        val firstColumnWidth = tableWidth - (columns - 1) * cellWidth
 
         val firstColumnRectangle = Rectangle(firstColumnWidth.toFloat(), cellHeight.toFloat())
         var firstColumnFontSize = PdfUtil.fitText(Font(bf), movesType[0], firstColumnRectangle, 10f, false, 1f)
 
+        // Decide the font size for the first column
         for (item in movesType) {
             firstColumnFontSize = Math.min(firstColumnFontSize, PdfUtil.fitText(Font(bf, firstColumnFontSize, Font.BOLD), item, firstColumnRectangle, 10f, false, 1f))
         }
         for (item in direction) {
             firstColumnFontSize = Math.min(firstColumnFontSize, PdfUtil.fitText(Font(bf, firstColumnFontSize), item, firstColumnRectangle, 10f, false, 1f))
         }
 
-        // Center the table
-        var maxFirstColumnWidth = 0f
-        var maxLastColumnWidth = 0f
+        var movesFontSize = 12f
+        val movesFont = Font(bf, movesFontSize)
 
+        var movesRectangle = Rectangle(cellWidth, cellHeight)
         for (i in movesType.indices) {
+            for (j in directionModifiers.indices) {
+                for (k in moves.indices) {
+                    movesFontSize = Math.min(movesFontSize, PdfUtil.fitText(movesFont, movesCell[i][j][k], movesRectangle, movesFontSize, false, 1f))
+                }
+            }
+        }
 
-            maxFirstColumnWidth = Math.max(maxFirstColumnWidth, bf.getWidthPoint(movesType[i], firstColumnFontSize))
+        val movesTypeFont = Font(bf, firstColumnFontSize, Font.BOLD)
 
-            var cell = PdfPCell(Phrase(movesType[i], Font(bf, firstColumnFontSize, Font.BOLD)))
-            cell.fixedHeight = cellHeight.toFloat()
-            cell.verticalAlignment = Element.ALIGN_MIDDLE
-            cell.horizontalAlignment = Element.ALIGN_RIGHT
-            cell.border = Rectangle.NO_BORDER
-            table.addCell(cell)
+        // Render moves
+        for (i in movesType.indices) {
```

Comment:
As far as I can tell, you are completely dropping the table mechanism in favor of hand-painted and pre-calculated rectangles.

This is not a good idea in my opinion. First, using a table makes sense when the end result should indeed look like a table. Second, it makes the code much harder to maintain since most of the drawing will likely break if you just change one small number somewhere.

Reply:
This is actually the main ideia. I tried to generate a sample including some moves and I found that do that using the table cells ended up with empty results. My test consisted in adding a new set of moves, namely, wide moves.